### ISSUE TITLE
Support HDR & wide color gamut on OSX

### ISSUE THREAD

**Issue by wjakob** [glfw/glfw#890]
Dear GLFW developers,

since OSX v10.11.2, applications now have the ability to request a wider color gamut than the default sRGB color space. Intensities can also be specified with a higher dynamic range (10 bit component values instead of the usual 8 bit). All of this works with OpenGL and can be accomplished simply by subscribing to certain Cocoa events -- see this page for the nitty gritty details: https://developer.apple.com/library/content/releasenotes/MacOSX/WhatsNewInOSX/Articles/MacOSX10_11_2.html.

After decades of 8 bit sRGB displays, it's exciting that the associated hardware is also becoming readily available (iMac & just-released MacBook Pro laptops). Unfortunately, GLFW applications have no way of requesting an OpenGL context with these features turned on, so they're still stuck in 8bit sRGB world.

It would be awesome if GLFW could introduce suitable hints to get access to this functionality. There is already a hint for specifying bit depths, so I suppose the API doesn't have to change for that part of the functionality.

    glfwWindowHint(GLFW_RED_BITS, 10);
    glfwWindowHint(GLFW_GREEN_BITS, 10);
    glfwWindowHint(GLFW_BLUE_BITS, 10);

For the other one (color spaces), a new flag would be required, e.g.

    glfwWindowHint(GLFW_COLOR_GAMUT, <argument>);

where "argument" is one of GLFW_DONT_CARE, GLFW_SRGB or GLFW_P3.

What do you think?

Best,
Wenzel

**Comment by elmindreda** [comment]
I'd love for GLFW to support this but I already have far too much to do for 3.3, so unless someone else implements it it'll have to wait a bit.

**Comment by elmindreda** [comment]
Looking into this. Do you have access to hardware to test this on?

**Comment by wjakob** [comment]
Yes -- I have access to both the iMac and a Macbook Pro with HDR & P3 gamut.

**Comment by elmindreda** [comment]
That's awesome! Then I'll _try_ to get this into the next release. I may need to ask you some silly questions, since the documentation seems pretty sparse.

**Comment by wjakob** [comment]
Hi @elmindreda,

please let me know if I can help in any way. I'm happy to test out experiments, etc. Is this still slated to go into 3.3?

Best,
Wenzel

**Comment by elmindreda** [comment]
@wjakob There isn't much you can do because this is already pretty much the perfect feature request. I just need to find time to read the relevant docs and think for a bit. I do expect this to go into 3.3.

**Comment by sagaceilo** [comment]
Hey ! I've just found that there is already HDR topic here, so I'll join. There is this presentation from nVidia: http://on-demand.gputechconf.com/siggraph/2016/presentation/sig1611-thomas-true-high-dynamic-range-rendering-displays.pdf that shows how to setup HDR on windpows (page 34). At least its some hint on Win platform. They are also use NVAPI to iterate HDR devices. Currently I'm working on HDR support in my game and I'm testing it on 4K HDR device (that have both hdr10 and Dolby Vision). If I run it successfuly I'll paste here some ideas and code.

**Comment by fsword7** [comment]
I researched and learned that only Quadro and FireGL cards allow HDR support.

**Comment by AlrikG** [comment]
What with Windows? This issue is tagged with macOS only. I have Windows 7, AMD r9 290x and a HDR10 Monitor which works perfectly with HDR games such as Battlefield V.
I also want the feature to create HDR Contexts with OpenGL for my game.

**Comment by janekb04** [comment]
As far as I know in Windows 10 creators update they added a universal way to create an hdr capable window.

**Comment by janekb04** [comment]
I did some searching and found a few resources on HDR. Unfortunately it doesn't seem very standardized, but at least it's something:

- On Windows, for DirectX, Microsoft provides some info [here](https://docs.microsoft.com/en-us/windows/win32/direct3darticles/high-dynamic-range)
- For Nvidia, they seem to provide their own APIs "until the OS integrates the support" [here](https://developer.nvidia.com/displaying-hdr-nuts-and-bolts)
- For Vulkan, I found the following extensions relevant: `VK_EXT_hdr_metadata`, `VK_AMD_display_native_hdr`

I'm not sure what could be done about OpenGL.

It seems that HDR support should probably be achievable, especially that other software such as Chromium and [VLC ](https://www.videolan.org/vlc/releases/3.0.0.html )have it. For Chromium I was able to find an issue regarding HDR [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1084689#c21).

I also found [this demo](https://github.com/kleinerm/jesse-cube) that shows off HDR support. I can't really test it, as I don't have access to an HDR monitor right now.

I hope this information helps.

**Comment by JoeHowse** [comment]
I am also very interested in HDR10 output but cross-platform support seems problematic.  @janekb04 has provided good information about some approaches that all include (at least) support for Windows 10.  I have started to do a bit of digging as well, and here is my understanding so far:

- Linux lacks support for HDR10 at the level of the display server (be it X.org or Wayland).  Since the display server does not support HDR10, neither does the window system or the desktop environment.  Although it is possible to bypass this HDR10-incompatible software stack under specific circumstances (e.g. in fullscreen using the Direct Display Mode feature of the open-source AMDVLK Vulkan drivers with a compatible AMD GPU), doing so seems too limiting for a window-creation library such as GLFW.
- The situation on the *BSDs is presumably no better than on Linux, since the *BSDs use ports of Linux display servers.
- On Windows 10, possible routes to HDR10 output include Direct3D 11, Direct3D 12, and Vulkan.  GLFW has no Direct3D back-end.  The Vulkan route might make the most sense for GLFW.
- On Windows 7-10, vendor-specific APIs from NVIDIA and AMD provide alternative routes to HDR10 output.  GLFW doesn't use such vendor-specific APIs, as far as I know.
- On Mac, Metal is a route to HDR10 output.  (For example, this route is used for [HDR10 support on Mac in NanoGUI](https://github.com/mitsuba-renderer/nanogui/commit/0146a88b2214cd5c5c29e6dfa8d3d3d0e9ab6d9d)).  However, GLFW has no Metal back-end.  I am not sure whether or not Vulkan --> MoltenVK --> Metal is a practical route to HDR10 output on Mac.

I'd be interested in anybody's further thoughts on possible approaches. Thanks!

**Comment by philipguin** [comment]
Would also appreciate this feature!

OpenGL doesn't appear to have an HDR extension yet, surprisingly, but unless they've stopped adding extensions altogether, I'd expect them to release one any day now... or maybe the OpenGL spec already allows this, instead deferring to platform-specific extensions to expose it?

I just got a 4K HDR display in the mail - would be happy to test any GLFW changes on my system!

**Comment by audaki** [comment]
FYI: Thanks to IMS212 there’s now a fork of GLFW that supports creating HDR windows, that works at least in Linux KDE 6 + HDR:

https://github.com/IMS212/glfw/tree/hdr

**Comment by IMS212** [comment]
Yep; The compositor just needs to be willing to accept an output. For windows, a DXGI swapchain would be best; it would fix many OGL and VK presentation issues as well as allowing HDR. I haven’t had a chance to look into it, though.

**Comment by audaki** [comment]
@elmindreda And for Windows I’ve copied some info by AnotherCommander, how they managed to output HDR in Windows with OpenGL but without having DXGI requirement (in the Oolite project):

https://github.com/IrisShaders/Iris/issues/2282#issuecomment-2155081329

**Comment by zezba9000** [comment]
@audaki Whats wrong with DXGI? This is fundamental to the Windows driver platform. Trying to avoid it seems more problematic?

**Comment by IMS212** [comment]
I have achieved scRGB HDR on Windows 11 natively using the approach described in #1907, while providing color primary information through a small DXGI function. It is incredibly primitive, but I can confirm it is functional.

**Comment by Lassebq** [comment]
Are there any implementations of wp-color-manager-v1 protocol yet?

**Comment by IMS212** [comment]
I tried it once, it seems much more complex for me than the old one. I’d be interested once there are more implementations than just the VK layer.

**Comment by rhjdvsgsgks** [comment]
> Are there any implementations of wp-color-manager-v1 protocol yet?

yes. https://github.com/rhjdvsgsgks/glfw/tree/colormanagement 

based on https://github.com/IMS212/glfw/commit/eb1f46529c6392cf3c6ac0dc8508c441ab3d8c8b , and almost copied from https://github.com/Zamundaaa/VK_hdr_layer/commit/09815451e063b9fcccaac9d370b3a7020c7e0c32 and https://github.com/libsdl-org/SDL/commit/fadb261b66ee5817008a813a9ea16feb9ec73b31

it is hard coded to bt 2020 pq. therefore only suitable for some of case
